### PR TITLE
Co-locate 'data' datasets with 'zone' datasets

### PR DESCRIPTION
**Pull request by smklein**
Fixes https://github.com/oxidecomputer/omicron/issues/4819

### CODE REVIEW

File: sled-agent/src/services.rs [hunk 1]
```diff
@@ -2932,11 +2929,33 @@ impl ServiceManager {
...
+                // service.
+                let data_pool = dataset.pool();
+                if !all_u2_pools.contains(&data_pool) {
+                    warn!(
+                        log,
+                        "zone dataset requested on a zpool which doesn't exist";
+                        "zone" => &name,
+                        "zpool" => %data_pool
+                    );
+                    return Err(Error::MissingDevice {
```
Comment (andrewjstone) [inline sled-agent/src/services.rs:2946]: Is there ever a case where we'd want to crate the zone filesystem even if the zpool dataset wasn't found? I can't think of one and believe this makes sense, just putting the question out there for completeness.

File: sled-agent/src/services.rs [hunk 2]
```diff
@@ -2932,11 +2929,33 @@ impl ServiceManager {
             //
             // This is (currently) intentional, as the zone filesystem should
             // be destroyed between reboots.
-            let mut rng = rand::thread_rng();
```
Comment (andrewjstone) [inline sled-agent/src/services.rs:2935]: Should probably remove the above comment and/or move part of it to the else clause below.

**Review by andrewjstone — APPROVED** [review]
LGTM, with one nit for the comment.

File: sled-agent/src/services.rs (hunk 1 above)
Comment (smklein) [reply sled-agent/src/services.rs:2946]: To be super pedantic:

> even if the zpool dataset wasn't found

This logic is not looking for a dataset, it's looking for a *zpool*. If the zpool doesn't exist, we should not expect to be able to provision any new datasets on it, for any reason.

The sled agent auto-formats new disks to have the necessary zpools for further allocations **before** those disks are made available to the control plane at large, so it should be necessary for us to process these pools before we could make this kinda allocation.

The only situation where I can imagine this causing a "new" error would be:

1. The sled agent boots, formats zpools, informs Nexus about them
2. The sled reboots, we haven't locally parsed which disks are up yet
3. Nexus tries sending a request to provision a new service to that sled
4. This logic runs, and we'd report "the U.2 you're asking for doesn't appear to exist"
5. If Nexus re-sent the request, it should work later, once the sled agent notices that the zpool exists

I *think* the value of the "early error in cases where the disk was truly removed" is more useful, but I could be persuaded to omit this error case, if you think it's a regression?

**Comment by smklein** [comment]
Manually testing on a local helios box:

# Before this PR

```bash
 $ zfs list -o name | grep cockroach                                            
oxp_14b4dc87-ab46-49fb-a4b4-d361ae214c03/cockroachdb
oxp_14b4dc87-ab46-49fb-a4b4-d361ae214c03/crypt/zone/oxz_cockroachdb_44c03fb8-0370-4ba0-9eb0-6dfb24da1afd
oxp_24b4dc87-ab46-49fb-a4b4-d361ae214c03/cockroachdb  # Not co-located with zone dataset
oxp_31bd71cd-4736-4a12-a387-9b74b050396f/cockroachdb
oxp_31bd71cd-4736-4a12-a387-9b74b050396f/crypt/zone/oxz_cockroachdb_f0c56182-9e76-4c81-afb7-47dfed16d051
oxp_616b26df-e62a-4c68-b506-f4a923d8aaf7/cockroachdb
oxp_616b26df-e62a-4c68-b506-f4a923d8aaf7/crypt/zone/oxz_cockroachdb_52e7363b-4409-4877-aa69-5bdece01dace
oxp_cd70d7f6-2354-4bf2-8012-55bf9eaf7930/cockroachdb  # Not co-located with zone dataset
oxp_e4b4dc87-ab46-49fb-a4b4-d361ae214c03/crypt/zone/oxz_cockroachdb_51b8666d-527f-4b4e-8269-4196eb48f27a
oxp_f4b4dc87-ab46-49fb-a4b4-d361ae214c03/crypt/zone/oxz_cockroachdb_74a8f98d-746e-43da-aa91-bc4740f1b9b1
```

# After this PR (all datasets co-located)

```bash
 $ zfs list -o name | grep cockroach
oxp_14b4dc87-ab46-49fb-a4b4-d361ae214c03/cockroachdb
oxp_14b4dc87-ab46-49fb-a4b4-d361ae214c03/crypt/zone/oxz_cockroachdb_5fa30186-8f64-47ef-9320-9d5c9fad6d11

oxp_24b4dc87-ab46-49fb-a4b4-d361ae214c03/cockroachdb
oxp_24b4dc87-ab46-49fb-a4b4-d361ae214c03/crypt/zone/oxz_cockroachdb_eb6efb03-0fec-4d1b-8260-7966d65b7a63

oxp_31bd71cd-4736-4a12-a387-9b74b050396f/cockroachdb
oxp_31bd71cd-4736-4a12-a387-9b74b050396f/crypt/zone/oxz_cockroachdb_4907da9c-6b8f-4203-a6b1-e331a7596a3a

oxp_616b26df-e62a-4c68-b506-f4a923d8aaf7/cockroachdb
oxp_616b26df-e62a-4c68-b506-f4a923d8aaf7/crypt/zone/oxz_cockroachdb_56cbfdb4-cd27-4eb0-84e9-f031a1817229

oxp_cd70d7f6-2354-4bf2-8012-55bf9eaf7930/cockroachdb
oxp_cd70d7f6-2354-4bf2-8012-55bf9eaf7930/crypt/zone/oxz_cockroachdb_fa5379b6-1519-4c7c-a191-930ad0c535cd
```

File: sled-agent/src/services.rs (hunk 1 above)
Comment (andrewjstone) [reply sled-agent/src/services.rs:2946]: Thanks for the details and pedantry. Always appreciated. :) I think the error is good to keep and in that case nexus should retry anyway I believe, whether with this sled or another. Eventually Inventory will handle the availability of disks and the planner will correct for this scenario I believe

File: sled-agent/src/services.rs (hunk 2 above)
Comment (smklein) [reply sled-agent/src/services.rs:2935]: Done in 61d94dce573b33d019dfe41023e4685285e410e1

**Pull request merged**